rfcs/20210519-single-pass-bnorm/README.md

```diff
@@ -0,0 +1,126 @@
+# Proposal to calculate mean & variance in batchnorm in single pass
+
+## Requester
+GPU Enabling team
+
+## Motivation
+Using new formula for calculating variance will result in up to 33% performance
+improvement in batchnorm fwd training operation. 
+
+Currently bnorm performance is far below projections. Projections are based on
```

Comment:
Are there any other optimizations which can be done to achieve the same performance or reduce existing gap?

Reply:
There is one method to reach projection, but it is not generic.
In bnorm, each channel of data has no influence on other channels, as if they were separate tensors. If:
- tensor is small enough that one channel worth of data can all fit into cache, and
- tensor layout allows loading a single channel at a time into cache/SRAM with full bandwidth - which generally means channel is not the innermost dimension (so NHWC is bad, NCHW is fine, etc)
then bnorm can load data for given channel once and do all 3 operations on it (mean, variance, normalization) while data stays in cache. In that case using new formula gives no benefit.

If either of the above points is not met (either C is innermost dimension and channels need to be loaded more-than-one at a time to sustain mem bandwidth, or there's too much data in each channel) then a weaker version of the above optimization may be used:
- for first pass (mean&var) load data for channel/set of channels starting from beginning of tensor; later loads will evict data loaded earlier, but the tail will stay in cache
- for second pass (normalization), load data backwards, starting from the end of tensor towards beginning. That way some data will be still available in cache and won't need to be loaded from main mem again.
In such scenario it again matters whether there are 2 or 3 passes and performance.

Reply:
@mgouicem : that's exactly what I tried to describe in my previous comment. Indeed this is the optimal way to do bnorm, but it is possible only if tensor, or its slice of channels, is small enough to fit into registers/SLM/L3$. If tensor is too big, data will spill out and need to be reloaded.

Reply:
Sorry I had misunderstood. Makes sense.